source/dub/generators/generator.d

```diff
@@ -143,144 +145,278 @@ class ProjectGenerator
+			if (tt != TargetType.none && tt != TargetType.sourceLibrary && ti.buildSettings.sourceFiles.empty) {
+				logWarn(`Configuration '%s' of package %s contains no source files. Please add {"targetType": "none"} to its package description to avoid building it.`,
+						ti.config, ti.pack.name);
+				tt = TargetType.none;
+			}
+			return tt;
 		}
 
-		shallowbs.targetType = tt;
-		bool generates_binary = tt != TargetType.sourceLibrary && tt != TargetType.none;
-		bool is_target = generates_binary || pack is m_project.rootPackage;
+		string[] mainSourceFiles;
+		bool[string] isBinaryTarget;
 
-		if (tt == TargetType.none) {
-			// ignore any build settings for targetType none (only dependencies will be processed)
-			shallowbs = BuildSettings.init;
-			shallowbs.targetType = TargetType.none;
-		}
+		foreach (ref ti; targets.byValue)
+		{
+			auto bs = &ti.buildSettings;
+			// determine the actual target type
+			bs.targetType = determineTargetType(ti);
 
-		// start to build up the build settings
-		BuildSettings buildsettings;
-		processVars(buildsettings, m_project, pack, shallowbs, true);
+			switch (bs.targetType)
+			{
+			case TargetType.none:
+				// ignore any build settings for targetType none (only dependencies will be processed)
+				*bs = BuildSettings.init;
+				bs.targetType = TargetType.none;
+				break;
+
+			case TargetType.executable:
+				break;
+
+			case TargetType.dynamicLibrary:
+				// set -fPIC for dynamic library builds
+				ti.buildSettings.addOptions(BuildOption.pic);
+				goto default;
+
+			default:
+				// remove any mainSourceFile from non-executable builds
+				if (bs.mainSourceFile.length) {
+					bs.sourceFiles = bs.sourceFiles.remove!(f => f == bs.mainSourceFile);
+					mainSourceFiles ~= bs.mainSourceFile;
+				}
+				break;
+			}
+			bool generatesBinary = bs.targetType != TargetType.sourceLibrary && bs.targetType != TargetType.none;
+			isBinaryTarget[ti.pack.name] = generatesBinary || ti.pack is rootPackage;
+		}
 
-		// remove any mainSourceFile from library builds
-		if (buildsettings.targetType != TargetType.executable && buildsettings.mainSourceFile.length) {
-			buildsettings.sourceFiles = buildsettings.sourceFiles.filter!(f => f != buildsettings.mainSourceFile)().array;
-			main_files ~= buildsettings.mainSourceFile;
+		// mark packages as visited (only used during upwards propagation)
+		void[0][Package] visited;
+
+		// collect all dependencies
+		void collectDependencies(Package pack, ref TargetInfo ti, TargetInfo[string] targets, size_t level = 0)
+		{
+			// use `visited` here as pkgs cannot depend on themselves
+			if (pack in visited)
+				return;
+			visited[pack] = typeof(visited[pack]).init;
+
+			auto bs = &ti.buildSettings;
+			if (isBinaryTarget[pack.name])
+				logDebug("%sConfiguring target %s (%s %s %s)", ' '.repeat(2 * level), pack.name, bs.targetType, bs.targetPath, bs.targetName);
+			else
+				logDebug("%sConfiguring non-binary target %s", ' '.repeat(2 * level), pack.name);
```

Comment:
Except potentially the root package, there is no such thing as a "non-binary target" - "non-target package" or similar would be more appropriate.